source/fem-dem/fluid_dynamics_vans_matrix_free_operators.cc

```diff
@@ -0,0 +1,499 @@
+
+  // Evaluate the grad-div stabilization constant
+  const unsigned int n_cells = this->matrix_free.n_cell_batches();
+  FECellIntegrator   integrator(this->matrix_free);
+
+  grad_div_gamma.reinit(n_cells, integrator.n_q_points);
+
+
+  for (unsigned int cell = 0; cell < n_cells; ++cell)
+    {
+      integrator.reinit(cell);
+      integrator.read_dof_values_plain(newton_step);
+
+      // Integrator must update the values since the velocity
+      // magnitude is used to calculate the grad-div stabilization constant.
+      integrator.evaluate(EvaluationFlags::values);
+      for (const auto q : integrator.quadrature_point_indices())
+        {
+          // Get the velocity magnitude to calculate the grad_div constant
+          VectorizedArray<number> u_mag_squared = 0;
+          for (unsigned int k = 0; k < dim; ++k)
+            u_mag_squared +=
+              Utilities::fixed_power<2>(integrator.get_value(q)[k]);
+          VectorizedArray<number> u = std::sqrt(u_mag_squared);
+          grad_div_gamma(cell, q) =
+            this->kinematic_viscosity + cfd_dem_parameters.cstar * u;
+        }
+    }
+}
+
+template <int dim, typename number>
+void
+VANSOperator<dim, number>::evaluate_void_fraction(
+  const VoidFractionBase<dim> &void_fraction_manager)
+{
+  AssertThrow(
+    void_fraction_manager.void_fraction_parameters->mode ==
+      Parameters::VoidFractionMode::function,
+    ExcMessage(
+      "The VANS matrix free solver only supports void fractions established using a function. Void fraction derived from functions will be developed in a future version of Lethe."));
+
+  this->timer.enter_subsection("operator::evaluate_void_fraction");
+
+  const unsigned int n_cells = this->matrix_free.n_cell_batches();
+  FECellIntegrator   integrator(this->matrix_free);
+
+  void_fraction.reinit(n_cells, integrator.n_q_points);
+  void_fraction_gradient.reinit(n_cells, integrator.n_q_points);
+
+  for (unsigned int cell = 0; cell < n_cells; ++cell)
+    {
+      integrator.reinit(cell);
+      for (const auto q : integrator.quadrature_point_indices())
+        {
+          const Point<dim, VectorizedArray<number>> point_batch =
+            integrator.quadrature_point(q);
+          // Temporary work around, extract the void fraction function from the
+          // void fraction manager
+          void_fraction(cell, q) = evaluate_function<dim, number>(
+            (void_fraction_manager.void_fraction_parameters->void_fraction),
+            point_batch);
+          void_fraction_gradient(cell, q) =
+            evaluate_function_gradient<dim, number>(
+              (void_fraction_manager.void_fraction_parameters->void_fraction),
+              point_batch);
+        }
+    }
+
+  this->timer.leave_subsection("operator::evaluate_void_fraction");
+}
+
+/**
+ * The expressions calculated in this cell integral are:
+ * (q,∇·δu) + (v, ε ∂t δu) + (v, ε (u·∇)δu) + (v, ε (δu·∇)u) - (∇·v, ε δp) +
```

Comment:
There are two terms that come from the continuity equation right? So this should be like this?:
```suggestion
 * (q,∇·(εδu)) + (v, ε ∂t δu) + (v, ε (u·∇)δu) + (v, ε (δu·∇)u) - (∇·v, ε δp) +
```
It might be better to write the two terms explicilty here so that it is clear...

Reply:
I did that and I added the missing pressure term also. Good catchj